crates/matrix-sdk-sqlite/src/crypto_store.rs

```diff
@@ -1488,37 +1499,52 @@ impl CryptoStore for SqliteCryptoStore {
         Ok(())
     }
 
+    #[instrument(skip(self))]
     async fn try_take_leased_lock(
         &self,
         lease_duration_ms: u32,
         key: &str,
         holder: &str,
-    ) -> Result<bool> {
+    ) -> Result<Option<CrossProcessLockGeneration>> {
+        let _timer = timer!("method");
+
         let key = key.to_owned();
         let holder = holder.to_owned();
 
-        let now_ts: u64 = MilliSecondsSinceUnixEpoch::now().get().into();
-        let expiration_ts = now_ts + lease_duration_ms as u64;
+        let now: u64 = MilliSecondsSinceUnixEpoch::now().get().into();
+        let expiration = now + lease_duration_ms as u64;
 
-        let num_touched = self
+        // Learn about the `excluded` keyword in https://sqlite.org/lang_upsert.html.
+        let generation = self
             .acquire()
             .await?
             .with_transaction(move |txn| {
-                txn.execute(
-                    "INSERT INTO lease_locks (key, holder, expiration_ts)
+                txn.query_row(
+                    "INSERT INTO lease_locks (key, holder, expiration)
                     VALUES (?1, ?2, ?3)
                     ON CONFLICT (key)
                     DO
-                        UPDATE SET holder = ?2, expiration_ts = ?3
-                        WHERE holder = ?2
-                        OR expiration_ts < ?4
-                ",
-                    (key, holder, expiration_ts, now_ts),
+                        UPDATE SET
+                            holder = excluded.holder,
+                            expiration = excluded.expiration,
+                            generation =
+                                CASE holder
+                                    WHEN excluded.holder THEN generation
+                                    ELSE generation + 1
+                                END
+                        WHERE
+                            holder = excluded.holder
+                            OR expiration < ?4
+                    RETURNING generation
```

Comment:
It feels a bit wrong to have this three times, but moving this to a common DB seems like it would cause other kind of trouble.

I don't have a better idea here :sweat_smile: